common/src/main/java/io/netty/util/internal/PlatformDependentCompanion.java

```
@@ -0,0 +1,88 @@
+   * {@link #checkSetMaxDirectMemoryWasCalled()} throws an {@link IllegalAccessException}
+   * if the max direct memory is not set.
+   */
+  public static final String JAVA_SYS_PROP_ENABLE_SIREN_CHECK = "siren.memory.direct.max.check.enable";
+
+  private static final InternalLogger logger = InternalLoggerFactory.getInstance(PlatformDependentCompanion.class);
+
+  private static final long MAX_DIRECT_MEMORY_NOT_SET = -2L;
+  private static final long MAX_DIRECT_MEMORY_USE_DEFAULT = -1L;
+
+  private PlatformDependentCompanion() {
+    throw new RuntimeException("By design this class cannot be instantiated.");
+  }
+
+  private static long maxDirectMemory = MAX_DIRECT_MEMORY_NOT_SET;
+
+  public static synchronized void setUseDefaultMaxDirectMemory() {
+    maxDirectMemory = MAX_DIRECT_MEMORY_USE_DEFAULT;
+  }
+
+  public static synchronized void setMaxDirectMemory(long m) {
+    if (m < 0) {
+      throw new IllegalArgumentException("Specified m must be greater or equal to 0");
+    }
+
+    maxDirectMemory = m;
+  }
+
+  public static synchronized long getMaxDirectMemory() {
+    return maxDirectMemory;
+  }
+
+  /**
+   * One can call this method to check that {@link #setMaxDirectMemory(long)} was called.
+   *
+   * @throws IllegalAccessException if {@link #setMaxDirectMemory(long)} was not called and Java system
+   *                                property {@link #JAVA_SYS_PROP_ENABLE_SIREN_CHECK} is set;
+   *                                otherwise it just returns.
+   */
+  public static synchronized void checkSetMaxDirectMemoryWasCalled() throws IllegalAccessException {
+    if (System.getProperty(JAVA_SYS_PROP_ENABLE_SIREN_CHECK, NO_CHECK_ENABLED).equals(NO_CHECK_ENABLED)) {
+      logger.warn("checkSetMaxDirectMemoryWasCalled() and is disabled. To enable this check, " +
+        "set Java system property '{}' to yes", JAVA_SYS_PROP_ENABLE_SIREN_CHECK);
+      return;
+    }
```

Why would we need this ?

First, this will always log a warning, unless the user explicitly add a jvm property at startup. This is not very user friendly.

Second, why would we enforce this in netty ? The use of the `PlatformDependentCompanion` is optional, it is a second mechanism that take precedence over the jvm property.

Enforcing the use of the `PlatformDependentCompanion` and the deprecation of the `JAVA_SYS_PROP_IO_NETTY_MAX_DIRECT_MEMORY` should be done on Federate side.

Imho, I would remove entirely the method `checkSetMaxDirectMemoryWasCalled` and the associated jvm property.

---

This allows a user of this optional class to make it mandatory as needed, which I think should be, i.e. mandatory, in Federate.
PD could be loaded before the Federate plugin, if the plugin is loaded using the same classloader as any other plugin. I would need to check if a plugin gets its own classloader or it uses the same for all plugins.

In Netty's unit tests I added initially more @Test to find out that each test were running using the same classloader so once PD was loaded it was initialized regardless of the PDC been set or not.

JAVA_SYS_PROP_IO_NETTY_MAX_DIRECT_MEMORY belongs to Netty, so I think the place to deprecate it is where it's created, unless you want just Federate not to use it in the future regardless of what Netty does, is that so?

---

> unless you want just Federate not to use it in the future regardless of what Netty does, is that so?

Yes, that was the idea.
In Federate, we could:
- display a deprecation warning when we detect that the `JAVA_SYS_PROP_IO_NETTY_MAX_DIRECT_MEMORY` is set, and explain that this is ignored
- check that the `PlatformDependent.directMemoryLimit()` returns the federate's root limit + 1. If that is not the case, then we could throw an exception so that we could catch cases where the `PlatformDependent` is not properly initialized.

---

Ok got it! Thx.